xbmc/addons/InputStream.cpp

```diff
@@ -0,0 +1,506 @@
+/*
+ *      Copyright (C) 2016 Team Kodi
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "InputStream.h"
+#include "utils/StringUtils.h"
+#include "utils/log.h"
+#include "cores/VideoPlayer/DVDDemuxers/DVDDemux.h"
+#include "utils/RegExp.h"
+
+namespace ADDON
+{
+
+std::unique_ptr<CInputStream> CInputStream::FromExtension(AddonProps props, const cp_extension_t* ext)
+{
+  std::string listitemprops = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@listitemprops");
+  std::string extensions = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@extension");
+  std::string name(ext->plugin->identifier);
+  return std::unique_ptr<CInputStream>(new CInputStream(std::move(props),
+                                                        std::move(name),
+                                                        std::move(listitemprops),
+                                                        std::move(extensions)));
+}
+
+CInputStream::CInputStream(AddonProps props, std::string name, std::string listitemprops, std::string extensions)
+: InputStreamDll(std::move(props))
+{
+  m_fileItemProps = StringUtils::Tokenize(listitemprops, "|");
+  for (auto &key : m_fileItemProps)
+  {
+    StringUtils::Trim(key);
+    key = name + "." + key;
+  }
+
+  m_extensionsList = StringUtils::Tokenize(extensions, "|");
+  for (auto &ext : m_extensionsList)
+  {
+    StringUtils::Trim(ext);
+  }
+}
+
+bool CInputStream::Supports(CFileItem &fileitem)
+{
+  std::string extension = URIUtils::GetExtension(fileitem.GetPath());
+  bool match = false;
+  for (auto &ext : m_extensionsList)
+  {
+    if (ext == extension)
+    {
+      match = true;
+      break;
+    }
+  }
+  if (!match)
+    return false;
+
+  if (!m_pStruct)
+    return true;
```

Comment:
- Supports() will be called with addon unloaded (m_pStruct=null)  (to check only the extension)
- if the extension matches, the dll is loaded and the supported URL prefixes are checked.

So returning true if m_pStruct=true is fine (th extensin was checked successfuly). I'll change it in my next commit.